### PR TITLE
Add query timeout for dns-rfc2136 plugin

### LINKED ISSUE #7084: dns-rfc2136 challenges run infinitely for invalid DNS servers
1. Set `dns_rfc2136_server` to something invalid, e.g. `dns_rfc2136_server = 2001:db8::`.
2. Try to obtain certificates.
3. Challenges run forever.

CTRL+C during a challenge:
```
[... usual]
Performing the following challenges:
dns-01 challenge for [...]
dns-01 challenge for [...]
Encountered exception:
Traceback (most recent call last):
  File "/usr/lib/python3.7/site-packages/certbot/auth_handler.py", line 75, in handle_authorizations
    resp = self._solve_challenges(aauthzrs)
  File "/usr/lib/python3.7/site-packages/certbot/auth_handler.py", line 139, in _solve_challenges
    resp = self.auth.perform(all_achalls)
  File "/usr/lib/python3.7/site-packages/certbot/plugins/dns_common.py", line 57, in perform
    self._perform(domain, validation_domain_name, validation)
  File "/usr/lib/python3.7/site-packages/certbot_dns_rfc2136/dns_rfc2136.py", line 76, in _perform
    self._get_rfc2136_client().add_txt_record(validation_name, validation, self.ttl)
  File "/usr/lib/python3.7/site-packages/certbot_dns_rfc2136/dns_rfc2136.py", line 112, in add_txt_record
    domain = self._find_domain(record_name)
  File "/usr/lib/python3.7/site-packages/certbot_dns_rfc2136/dns_rfc2136.py", line 186, in _find_domain
    if self._query_soa(guess):
  File "/usr/lib/python3.7/site-packages/certbot_dns_rfc2136/dns_rfc2136.py", line 209, in _query_soa
    response = dns.query.udp(request, self.server, port=self.port)
  File "/usr/lib/python3.7/site-packages/dns/query.py", line 240, in udp
    _wait_for_readable(s, expiration)
  File "/usr/lib/python3.7/site-packages/dns/query.py", line 157, in _wait_for_readable
    _wait_for(s, True, False, True, expiration)
  File "/usr/lib/python3.7/site-packages/dns/query.py", line 131, in _wait_for
    if not _polling_backend(fd, readable, writable, error, timeout):
  File "/usr/lib/python3.7/site-packages/dns/query.py", line 87, in _poll_for
    event_list = pollable.poll()
KeyboardInterrupt
```

**Expected:**  
Certbot should set a reasonable query timeout ([this is supported by `dnspython`](https://stackoverflow.com/questions/8989457/dnspython-setting-query-timeout-lifetime)) and alert the user of an issue with the DNS server.

### PR DESCRIPTION
**Pull request by osirisinferi**
Fixes #7084 IMO.

I've used `acme/acme/client.py` as an example (OK, *fine*, I've stolen the code from there! You got me.. :smile: ) and a network timeout of 45 seconds is hardcoded there. I didn't have any reason to change the hardcoded timeout (these DNS queries are network thingies too, just like in the `acme` library, right?), so I left it at 45 seconds.

Note: this PR doesn't handle retrying a failed query as suggested in the issue this PR hopes to fix.

TODO: terminating the whole authorization deal when encountering an exception. As of now, the certbot DNS auth train seems to continue despite a heck of a lot of exceptions.. I'm pretty sure that's not the expected behaviour? Shouldn't the plugin quit everything it was doing when `query_soa()` already fails?

### CODE REVIEW

**Comment by ohemorange** [comment]
The code in this PR looks good, but I noticed your TODO -- would you mind tracing through why the exception doesn't stop the auth train?

**Comment by osirisinferi** [comment]
> The code in this PR looks good, but I noticed your TODO -- would you mind tracing through why the exception doesn't stop the auth train?

I would, but I have some other thoughts about that too: what if *one* of *multiple* challenges fails? Then certbot would still need to clean up the challenges which *did* succeed, right? We can't just quit half way multiple challenges because *one* of them has a timeout (possibly b/c of a temporary network glitch or something).

Because of the above, I thought it needed some extra coding, like have some kind of "list" of challenges which *did* succeed and after a time out, clean up only the challenges on that "succeeded challenges" list. That *could* be out of my coding-league though..

**Comment by ohemorange** [comment]
> I would, but I have some other thoughts about that too: what if one of multiple challenges fails? Then certbot would still need to clean up the challenges which did succeed, right? We can't just quit half way multiple challenges because one of them has a timeout (possibly b/c of a temporary network glitch, MTU issue, strange firewall behaviour or something).

That's a good point; we should probably only fail after *all* challenges have failed. Do we really try multiple challenges simultaneously, not sequentially? have you found the code that handles that yet?

**Comment by ohemorange** [comment]
Hey, just checking in on this PR. Any updates?

**Comment by osirisinferi** [comment]
Currently not, but I'm planning on trying to look at what you suggested earlier the upcoming days.

**Comment by ohemorange** [comment]
Great, thanks for the update!

**Comment by osirisinferi** [comment]
> Do we really try multiple challenges simultaneously, not sequentially?

Currently, challenges are performed as followed which looks semi-simultaneously to me:

```
dns-01 challenge for working-dns-challenge.example.com
dns-01 challenge for forcefully-timedout-dns-challenge.example.com
Adding TXT record
Query_Soa
Query_Soa
Adding TXT record
Query_Soa
Query_Soa
Query_Soa
Cleaning up challenges
Removing TXT record..
Query_Soa
Query_Soa
Removing TXT record..
Query_Soa
Query_Soa
Query_Soa
```

If the second challenge fails somewhere during the process due to a time out, certbot continues like nothing has happened. This is partly *good* because it means no lingering TXT records for records which were added succesfully. However, it also means possibly unnecessary "Removing TXT record" attempts if the **adding** of the TXT record was unsuccesful anyway. Although I'm not sure what happens if the (TCP!) connection to the DNS server would fail *after* the client has send the "please add the TXT record". This *might* happen if for some reason the ACK TCP package would be dropped for some reason: I think the client would give a time out exception, but the DNS server actually receiving the "Add TXT record" command might have added the record anyway. So I'm personally OK with just leaving the "cleanup" active despite the majority of failed add-record-attempts wouldn't need one.

**Review by ohemorange — APPROVED** [review]
You know, this code is correct, and any issues with error handling are really going to be a separate issue, so let's just get this in. @osirisinferi would you mind making a new issue to look into making sure that we're handling cleanup for multiple challenges correctly?

File: certbot/CHANGELOG.md [hunk 1]
```diff
@@ -20,6 +20,7 @@ Certbot adheres to [Semantic Versioning](https://semver.org/).
 * The `acme` library can now tell the ACME server to clear contact information by passing an empty
   `tuple` to the `contact` field of a `Registration` message. 
 * Fixed the `*** stack smashing detected ***` error in the Certbot snap on some systems.
+* Added timeout to DNS query function calls for dns-rfc2136 plugin.
```
Comment (ohemorange) [inline certbot/CHANGELOG.md:23]: Oh actually, this should be moved up to the new section after merging master.

**Review by ohemorange — CHANGES_REQUESTED** [review]

File: certbot/CHANGELOG.md (hunk 1 above)
Comment (osirisinferi) [reply certbot/CHANGELOG.md:23]: I'll try to, hopefully without messing up the branch.

**Review by ohemorange — APPROVED** [review]

**Pull request merged**